### DISCUSSION TITLE
One calculator to rule them all?

### DISCUSSION THREAD

**Discussion by bsidhom** [sharkdp/numbat#368] [Ideas]
I actually happened upon Numbat because I was looking for something akin to [calc](https://github.com/lcn2/calc), an arbitrary precision calculator, which I could use on the web (on my phone). Better yet if it could merge that functionality with that of [Units](https://www.gnu.org/software/units/)! I remembered a project called Insect which billed itself as "high precision" (which I think I interpreted as "arbitrary precision"), which brought me here. When I'm at my computer, I use calc for basically all one-off computations. It's great because it's fast, has runtime-configurable arbitrary precision, is scriptable in a C-like language, and includes a plethora of [special functions and utility functions](https://github.com/lcn2/calc/tree/master/cal) out of the box.

I would love to see Numbat fill this void. All it's really missing at the moment is support for [arbitrary precision arithmetic](https://github.com/sharkdp/numbat/issues/4) and [better function ergonomics](https://github.com/sharkdp/numbat/discussions/367) to make user-contributed scripting more feasible. It would also be nice if it included useful control flow mechanics such as looping (though if tail call optimization is implemented, that's less of a big deal).

In the high-precision issue I did propose a finite-precision solution (based on quad-doubles) mostly as a stop-gap, but arbitrary precision could make this a very powerful tool.

What do people think about this vision? Am I barking up the wrong tree? Is there another WASM calculator project that I've missed? I did play around with QuickJS/QJSCalc for this purpose. It's quick, scriptable in JS, uses arbitrary precision, and is under a very permissive license, but it lacks robust typing and a unit system altogether. I feel like Numbat with arbitrary precision and "contrib" modules could tick _all_ the boxes.

**Comment by sharkdp** [comment]
> I actually happened upon Numbat because I was looking for something akin to [calc](https://github.com/lcn2/calc), an arbitrary precision calculator, which I could use on the web (on my phone). Better yet if it could merge that functionality with that of [Units](https://www.gnu.org/software/units/)! I remembered a project called Insect which billed itself as "high precision" (which I think I interpreted as "arbitrary precision"), which brought me here.

Welcome!

> has runtime-configurable arbitrary precision

> arbitrary precision could make this a very powerful tool

I'd be interested in some real world use cases for arbitrary precision computations. I would *definitely* like to have big integers in Numbat so we can perform computations with arbitrary large integers. But so far, I think I'm not 100% convinced that we definitely need arbitrary precision decimals. I mean, I would take them if we can get them from somewhere. But so far, I did not find a Rust crate that serves our purposes.

> includes a plethora of [special functions and utility functions](https://github.com/lcn2/calc/tree/master/cal) out of the box.

Mostly out of curiosity: which of these functions do you consider essential for a calculator like Numbat?

> [better function ergonomics](https://github.com/sharkdp/numbat/discussions/367) to make user-contributed scripting more feasible

We'll work on that for sure. Let's discuss details in the other ticket.

> if it included useful control flow mechanics such as looping (though if tail call optimization is implemented, that's less of a big deal).

Up until now, Numbat feels more like a functional programming language to me. So I would probably add lambdas, lists, and then `map`/`filter`/etc. at some point. Even if we don't have TCO in the beginning, we could at least make things like `map` and `filter` fast by implementing them in Rust (instead of Numbat).

> an arbitrary precision calculator, which I could use on the web (on my phone).
> Better yet if it could merge that functionality with that of [Units](https://www.gnu.org/software/units/)

> What do people think about this vision?

I mean, I'm biased — but I certainly share that vision for Numbat (maybe except for the *arbitrary* precision part, which I'm not yet sure about, see above).

> an arbitrary precision calculator, which I could use on the web (on my phone). Better yet if it could merge that functionality with that of [Units](https://www.gnu.org/software/units/)

I don't know. You can check out #19. And let us know in case you find something which better suits your needs :smile: 

> I feel like Numbat with arbitrary precision and "contrib" modules could tick _all_ the boxes.

:star_struck:

**Reply by bsidhom** [reply]
> Mostly out of curiosity: which of these functions do you consider essential for a calculator like Numbat?

The most important to _me personally_ are the complete and incomplete gamma and beta functions and Lambert-W. For the latter, it would be great to have the full multi-valued function but I’ve only ever used the principal branch in practice.

These typically arise in probability/statistical problems that I’m working on and want to deal with high precision over large numbers. 128-bit floats are almost certainly enough, but I’ve used more than that before, typically out of curiosity. When I don’t have access to AP, I can typically work around this by using log-space, but often times the actual values are discrete. Gamma and beta come up much more often than Lambert-W, but the latter is still very useful for solving physical systems and recurrences.

**Reply by bsidhom** [reply]
Also just noticed your note about big _integers_ rather than decimals. That is precisely what I use when I need AP (usually in solving combinatorial problems). In other words, I don’t (yet) have any concrete use cases for AP decimals. 🙂

**Reply by sharkdp** [reply]
> The most important to me personally are the complete and incomplete gamma and beta functions and Lambert-W. For the latter, it would be great to have the full multi-valued function but I’ve only ever used the principal branch in practice.

We currently only have support for the complete gamma function via [libm](https://docs.rs/libm/latest/libm/index.html). It does provide a few other functions, but none of the ones mentioned by you.

We could look into depending on [statrs](https://docs.rs/statrs/latest/statrs/function/index.html), which seems to have support for incomplete gamma and beta functions, but not Lambert-W. I haven't looked into any alternatives so far. But in principle, if we find a implementation in Rust, it should be easy to add new FFI functions to Numbat (`exp` or `gamma` can serve as examples).

**Reply by bsidhom** [reply]
I didn't realize that libm was pure rust. That's good, though it looks like it also specializes for f32 and f64, so support for [qd](https://github.com/Barandis/qd) (if we were to go that route) might not work. (Same applies to statrs.) I think that's a longer term investigation in any case, so I wouldn't necessarily want to block on that.

What I was originally picturing, though, was something more similar to what [calc](https://github.com/lcn2/calc) does. Specifically, it offers high-performance arbitrary precision arithmetic and elementary function implementations as builtins and then ships with `.cal` libraries (equivalent to user-defined functions in `.nbt` modules) for a large suite of special functions. The nice thing about that is that it's very extensible and easy to add more implementations (assuming there are well known algorithms for computing to AP). As discussed above, AP for floats is likely unnecessary though, so I'm not sure how big of a deal that would be. It would be _nice_ to have support for something like f128+, but, again, this requires more thought.

**Comment by bsidhom** [comment]
> Up until now, Numbat feels more like a functional programming language to me. So I would probably add lambdas, lists, and then map/filter/etc. at some point. Even if we don't have TCO in the beginning, we could at least make things like map and filter fast by implementing them in Rust (instead of Numbat).

That sounds excellent!

**Comment by hunter-gatherer8** [comment]
> But so far, I think I'm not 100% convinced that we definitely need arbitrary precision decimals

Just came here to echo the sentiment that without arbitrary precision numbers (mostly I think about integers, though) numbat really cannot serve as a general-purpose calculator, which I'd really love it to be, since it already covers most of use cases I actually need ergonomic everyday calculator for (which is mostly stupid simple stuff, like unit conversion or basic operations in non-SI units, datetime operations, etc).

I mean:
```
>>> 1000!/(1000-1)!

    = NaN 
```

So all of combinatorics and probability (which, for me is probably the third most common use-case for day-to-day calculator) goes out of the window just like that.

For some reference, [kalker](https://github.com/PaddiM8/kalker) does this just fine, and it also is written in Rust and compiles to wasm, so maybe it's relevant. I didn't look into the actual implementation though.

_Edit:_ …However, now that I think of it, this probably has more with symbolic simplifications than with big integers. I mean, yes, kalker will tell me an exact value of 100!, which numbat cannot (and this is due to large integers/arbitrary precision), but it also will tell me that log(1 000 000!, 10)
 is 5 565 708.9171867, and of course, it couldn't actually compute 1 000 000! so fast… I think?

So, is it possible to make numbat more capable of tricks like that?

**Reply by sharkdp** [reply]
Let's continue the discussion around arbitrary precision floats in #4.

> For some reference, [kalker](https://github.com/PaddiM8/kalker) does this just fine

Really? It yields a result for 100!, but for 1000! it yields "not defined"

![image](https://github.com/sharkdp/numbat/assets/4209276/bd12ba6f-e3c4-480c-b90f-493f2d4a9802)

> but it also will tell me that log(1 000 000!, 10) is 5 565 708.9171867,

Oh, I see that the CLI version can actually compute `log(1000000!, 10)`. So they seem to use a different math library for the web and natively.

> this probably has more with symbolic simplifications than with big integers

I don't think so. For slightly larger numbers, it will still say "Not defined".

> So, is it possible to make numbat more capable of tricks like that?

We're most probably not going to add CAS-like features to Numbat. It will stay a *numerical* calculator. That said, I definitely want to support big integers. But symbolic reasoning and simplification is probably not going to happen. It's easy to do for a few hand-selected cases, but very hard as a general problem.

**Comment by hunter-gatherer8** [comment]
Yes, I've tried (and am using in general) only CLI version, so wasn't aware that web-version fails to do that.

> I don't think so. For slightly larger numbers, it will still say "Not defined".

Maybe. I still haven't taken a look at the source code. Indeed, kalker does it fine for 10 000 000, but thinks that log (100 000 000!, 10) = ∞, which isn't actually _that_ big. So probably you are right. That being said, my concern is exclusively practical, so if this can be done purely by employing big int math — great. If this can be done with CAS, but only for a particular hand-selected case and not in general — also great. The motivation for the question is the desire to find a "calculator to rule them all" as the OP has put it, and while it's surely would be nice for it to handle even these huge examples, honestly, I've only tried it because of this thread. I've never really had _a practical_ reason to estimate 100 000 000!. Finding something of order C(1000, 5) on the other hand — that happens all the time, and this is very cumbersome to do unless the calculator can handle 1000! (or just has C(n,k) implemented as a built-in).

(...Or as a separately implemented user function, which also is perfectly doable, to be fair, but somehow it occurred to me only just now — probably because I never actually think about C(n,k) as a separate function, but instead as n!/k!(n-k)!, so that's what I always type into a calculator...)
